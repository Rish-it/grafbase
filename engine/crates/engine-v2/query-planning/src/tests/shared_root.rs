use crate::{
    tests::{read_schema, strdiff, TestOperation},
    OperationGraph,
};

const SCHEMA: &str = r###"
enum join__Graph {
  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
}

type Product
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
  category: String @join__field(graph: CATEGORY)
  name: String @join__field(graph: NAME)
  price: Float @join__field(graph: PRICE)
}

type Query
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  product: Product
  products: [Product]
}
"###;

#[test]
fn all_fields() {
    let schema = read_schema(SCHEMA);
    let mut operation = TestOperation::bind(
        &schema,
        r#"
        query {
          products {
            price
            category
            id
            name
          }
        }
    "#,
    );

    let mut graph = OperationGraph::new(&schema, &mut operation);
    insta::assert_snapshot!("all_fields-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());

    let before = graph.to_dot_graph();

    graph.prune_resolvers_not_leading_to_any_scalar_node();
    insta::assert_snapshot!(strdiff(&before, &graph.to_dot_graph()), @"");
}

#[test]
fn single_field() {
    let schema = read_schema(SCHEMA);
    let mut operation = TestOperation::bind(
        &schema,
        r#"
        query {
          products {
            price
          }
        }
    "#,
    );

    let mut graph = OperationGraph::new(&schema, &mut operation);
    insta::assert_snapshot!("single_field-graph", graph.to_dot_graph(), &graph.to_pretty_dot_graph());

    let before = graph.to_dot_graph();

    graph.prune_resolvers_not_leading_to_any_scalar_node();
    insta::assert_snapshot!(strdiff(&before, &graph.to_dot_graph()), @r##"
    -    3 [ Root#category]
    -    4 [ products@Root#category]
    -    5 [ Root#name]
    -    6 [ products@Root#name]
    -    0 -> 3 [ label = "CreateChildResolver(1)" ]
    -    0 -> 3 [ label = "HasChildResolver" ]
    -    3 -> 4 [ label = "CanProvide(0)" ]
    -    4 -> 2 [ label = "Provides" ]
    -    0 -> 5 [ label = "CreateChildResolver(1)" ]
    -    0 -> 5 [ label = "HasChildResolver" ]
    -    5 -> 6 [ label = "CanProvide(0)" ]
    -    6 -> 2 [ label = "Provides" ]
    "##);
}
