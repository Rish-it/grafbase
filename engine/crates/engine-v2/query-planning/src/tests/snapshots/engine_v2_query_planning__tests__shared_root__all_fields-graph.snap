---
source: engine/crates/engine-v2/query-planning/src/tests/shared_root.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"price\" ]\n    2 [ label = \"category\" ]\n    3 [ label = \"id\" ]\n    4 [ label = \"name\" ]\n    5 [ label = \"products\" ]\n    6 [ label = \"Root#category\", shape=box, color=blue ]\n    7 [ label = \"products@Root#category\", shape=box, style=dashed, color=blue ]\n    8 [ label = \"Root#name\", shape=box, color=blue ]\n    9 [ label = \"products@Root#name\", shape=box, style=dashed, color=blue ]\n    10 [ label = \"Root#price\", shape=box, color=blue ]\n    11 [ label = \"products@Root#price\", shape=box, style=dashed, color=blue ]\n    12 [ label = \"id@Root#price\", shape=box, style=dashed, color=blue ]\n    13 [ label = \"price@Root#price\", shape=box, style=dashed, color=blue ]\n    14 [ label = \"name@Root#name\", shape=box, style=dashed, color=blue ]\n    15 [ label = \"id@Root#name\", shape=box, style=dashed, color=blue ]\n    16 [ label = \"id@Root#category\", shape=box, style=dashed, color=blue ]\n    17 [ label = \"category@Root#category\", shape=box, style=dashed, color=blue ]\n    0 -> 5 [ label = \"\" ]\n    0 -> 6 [ label = <<b>1</b>>, color=blue,fontcolor=blue ]\n    0 -> 6 [ label = \"\", style=dashed ]\n    6 -> 7 [ label = \"\", color=blue,fontcolor=blue ]\n    7 -> 5 [ label = \"\", color=turquoise ]\n    0 -> 8 [ label = <<b>1</b>>, color=blue,fontcolor=blue ]\n    0 -> 8 [ label = \"\", style=dashed ]\n    8 -> 9 [ label = \"\", color=blue,fontcolor=blue ]\n    9 -> 5 [ label = \"\", color=turquoise ]\n    0 -> 10 [ label = <<b>1</b>>, color=blue,fontcolor=blue ]\n    0 -> 10 [ label = \"\", style=dashed ]\n    10 -> 11 [ label = \"\", color=blue,fontcolor=blue ]\n    11 -> 5 [ label = \"\", color=turquoise ]\n    5 -> 4 [ label = \"\" ]\n    5 -> 3 [ label = \"\" ]\n    11 -> 12 [ label = \"\", color=blue,fontcolor=blue ]\n    12 -> 3 [ label = \"\", color=turquoise ]\n    5 -> 2 [ label = \"\" ]\n    5 -> 1 [ label = \"\" ]\n    11 -> 13 [ label = \"\", color=blue,fontcolor=blue ]\n    13 -> 1 [ label = \"\", color=turquoise ]\n    9 -> 14 [ label = \"\", color=blue,fontcolor=blue ]\n    14 -> 4 [ label = \"\", color=turquoise ]\n    9 -> 15 [ label = \"\", color=blue,fontcolor=blue ]\n    15 -> 3 [ label = \"\", color=turquoise ]\n    7 -> 16 [ label = \"\", color=blue,fontcolor=blue ]\n    16 -> 3 [ label = \"\", color=turquoise ]\n    7 -> 17 [ label = \"\", color=blue,fontcolor=blue ]\n    17 -> 2 [ label = \"\", color=turquoise ]\n}\n"
---
digraph {
    0 [ root]
    1 [ price]
    2 [ category]
    3 [ id]
    4 [ name]
    5 [ products]
    6 [ Root#category]
    7 [ products@Root#category]
    8 [ Root#name]
    9 [ products@Root#name]
    10 [ Root#price]
    11 [ products@Root#price]
    12 [ id@Root#price]
    13 [ price@Root#price]
    14 [ name@Root#name]
    15 [ id@Root#name]
    16 [ id@Root#category]
    17 [ category@Root#category]
    0 -> 5 [ label = "Field" ]
    0 -> 6 [ label = "CreateChildResolver(1)" ]
    0 -> 6 [ label = "HasChildResolver" ]
    6 -> 7 [ label = "CanProvide(0)" ]
    7 -> 5 [ label = "Provides" ]
    0 -> 8 [ label = "CreateChildResolver(1)" ]
    0 -> 8 [ label = "HasChildResolver" ]
    8 -> 9 [ label = "CanProvide(0)" ]
    9 -> 5 [ label = "Provides" ]
    0 -> 10 [ label = "CreateChildResolver(1)" ]
    0 -> 10 [ label = "HasChildResolver" ]
    10 -> 11 [ label = "CanProvide(0)" ]
    11 -> 5 [ label = "Provides" ]
    5 -> 4 [ label = "Field" ]
    5 -> 3 [ label = "Field" ]
    11 -> 12 [ label = "CanProvide(0)" ]
    12 -> 3 [ label = "Provides" ]
    5 -> 2 [ label = "Field" ]
    5 -> 1 [ label = "Field" ]
    11 -> 13 [ label = "CanProvide(0)" ]
    13 -> 1 [ label = "Provides" ]
    9 -> 14 [ label = "CanProvide(0)" ]
    14 -> 4 [ label = "Provides" ]
    9 -> 15 [ label = "CanProvide(0)" ]
    15 -> 3 [ label = "Provides" ]
    7 -> 16 [ label = "CanProvide(0)" ]
    16 -> 3 [ label = "Provides" ]
    7 -> 17 [ label = "CanProvide(0)" ]
    17 -> 2 [ label = "Provides" ]
}
